tests/Facility.CodeGen.CSharp.UnitTests/ValidationTests.cs

```diff
@@ -0,0 +1,85 @@
+using System.Linq;
+using Facility.Definition;
+using Facility.Definition.CodeGen;
+using Facility.Definition.Fsd;
+using NUnit.Framework;
+
+namespace Facility.CodeGen.CSharp.UnitTests
+{
+	public class ValidationTests
+	{
+		[Test]
+		public void RequestDtoGeneratesValidation()
+		{
+			var definition = "[csharp] service TestApi { method do { [validate(value: 0..1)] normalized: double; }: {} }";
+			var requestDtoFile = GetGeneratedFile(definition, "DoRequestDto.g.cs");
+
+			StringAssert.Contains("public override bool Validate", requestDtoFile.Text);
+		}
+
+		[Test]
+		public void DataDtoGeneratesRangeValidation()
+		{
+			var definition = "[csharp] service TestApi { method do { data: Data; }: {} data Data { [validate(value: 0..1)] normalized: double; } }";
+			var requestDtoFile = GetGeneratedFile(definition, "DataDto.g.cs");
+
+			StringAssert.Contains("public override bool Validate", requestDtoFile.Text);
+		}
+
+		[Test]
+		public void GeneratesNumericValidation()
+		{
+			var definition = "[csharp] service TestApi { method do { [validate(value: 0..1)] normalized: double; }: {} }";
+			var requestDtoFile = GetGeneratedFile(definition, "DoRequestDto.g.cs");
+
+			StringAssert.Contains("if (Normalized < 0D)", requestDtoFile.Text);
+			StringAssert.Contains("if (Normalized > 1D)", requestDtoFile.Text);
+		}
+
+		[Test]
+		public void GeneratesEnumValidation()
+		{
+			var definition = "[csharp] service TestApi { method do { [validate] data: Data; }: {} enum Data { nothing, something } }";
+			var requestDtoFile = GetGeneratedFile(definition, "DoRequestDto.g.cs");
+
+			StringAssert.Contains("public override bool Validate", requestDtoFile.Text);
+			StringAssert.Contains("if (!Data.IsDefined())", requestDtoFile.Text);
+		}
+
+		[Test]
+		public void GeneratesStringLength()
+		{
+			var definition = "[csharp] service TestApi { method do { [validate(length: 10..)] password: string; }: {} }";
+			var requestDtoFile = GetGeneratedFile(definition, "DoRequestDto.g.cs");
+
+			StringAssert.Contains("if (Password.Length < 10)", requestDtoFile.Text);
+		}
+
+		[Test]
+		public void GeneratesStringPattern()
+		{
+			var definition = @"[csharp] service TestApi { method do { [validate(regex: ""[0-9]{4}"")] pin: string; }: {} }";
```

Comment:
```suggestion
			var definition = @"[csharp] service TestApi { method do { [validate(regex: ""^[0-9]{4}$"")] pin: string; }: {} }";
```
I'd add anchors so that the test doesn't cause confusion on how validation works.